api/envoy/type/matcher/v3/header.proto

```diff
@@ -0,0 +1,57 @@
+syntax = "proto3";
```

Comment:
I'm curious if we can land https://github.com/envoyproxy/envoy/pull/13365 and make the protos there the basis for the https://github.com/envoyproxy/envoy/pull/12938 solution, or whether we should go with this PR?

@snowp any thoughts? CC @envoyproxy/api-shepherds

Reply:
IMO it's going to be a while before we can sort out how to use the common matching structure as part of routing so I would be fine with doing this for now.

Reply:
Thanks @mattklein123 .

@htuch Since this PR no longer has any refactoring, only copying, I rebased the contents for this PR in the original PR.

I am almost done with integrating MatchPredicate there. Please start taking a look. Thanks!
https://github.com/envoyproxy/envoy/pull/12938